### PR TITLE
Add phone verification

### LINKED ISSUE #11339: Add Phone Verification for new Users
In order to reduce the ever increasing amount of abuse on Gitpod, we'd like to make it hard for such people to get onto Gitpod. The hypothesis is that phone number verification would add an extra step that can reduce the amount.

The verification should happen before after signup but before they can start their first workspace. We could further reduce the friction for good users by whitelisting based on certain criteria (e.g. age of GH or GL account). 

The verification should only be active on SaaS and we should make use of an external service such as Twillio's Verify.

See **[design specs](https://github.com/gitpod-io/gitpod/issues/11339#issuecomment-1183387915)** below. 🥯

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
This PR adds phone number verification before untrusted users can start a workspace.

A user is considered trusted if
 - Their GitHub account is older than 30 days, or
 - Their GitLab account `confirmed_at` event is older than 30 days, or
 - Their Gitpod account has been created before 2022-08-24 (i.e. existing users)

## Related Issue(s)
Fixes #11339 

See also https://github.com/gitpod-io/ops/pull/4149

## How to test
*Test 1*
 - Create a fresh GitHub account and sign in (use admin to unblock). 
 - You should need to do phone number verification before you can start your first workspace.
 - Once the number is verified you should be able to work as usual.

*Test 2*
 - Use a GitHub account older than 30 days.
 - You should be able to start a workspace without verification.

*Test 1 ++*
 - delete the `lastVerificationTimestamp` in your user's additional Data.
 - if the account had started a workspace before, you should not need to verify
 - if the account doesn't have any workspaces, you should be asked to verify

## Release Notes

```release-note
Added phone number verification in SaaS, to help mitigate abuse.
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:164]: **issue**: Generally, placeholders are problematic for multiple reasons. Some older browsers don't support them[[1](https://caniuse.com/input-placeholder)] although this can be considered minor, some browsers hide them on focus, it's harder for users to check if their input complies with the placeholder hint, have lower contrast[[2](https://usecontra.st/a8a29e/fff)] which is easy to fail for color contrast accessibility, some screen readers don’t recognise them, and much more.

**suggestion**: Instead, we could use a help text below or above the input when possible. In this case, we can safely drop this placeholder text and rely on the input label and the instructions above for users to understand what action they need to take. That's the primary reason why the design specs in https://github.com/gitpod-io/gitpod/issues/11339#issuecomment-1183387915 do not include any placeholders.

```suggestion
```

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:11]: **question**: Unless there's a strong preference over this library, given the styling issues of the input, shall we consider switching to a different library like [`intl-tel-input`](https://github.com/jackocnr/intl-tel-input) which also happens to be sponsored by Twilio? Seems more popular and flexible, no?

> 270 vs. 6.4K ⭐

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:55]: **nitpick**: Does it help us achieve our goal (abuse)? I'd rather go with a neutral tone but any option could suffice. Feel free to ignore this nitpick. 🙂 

```suggestion
                    To use Gitpod you'll need to validate your account with a mobile phone number. This is required to
```

**Review by gtsiolis — COMMENTED** [review]
Left some early feedback below. 🥯

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:164]: We only support the current version of chrome, firefox and safari.

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:11]: It is the react version of that library.

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:159]: Please try again. The authtoken was invalidated.

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:184]: Regarding https://github.com/gitpod-io/gitpod/pull/12258#discussion_r952581084:

> We only support the current version of chrome, firefox and safari.

Agree, makes sense. 🤝 Maybe worth being more specific about this also in [our docs](https://www.gitpod.io/docs/self-hosted/helm-deprecated/requirements#supported-web-browsers). Opened https://github.com/gitpod-io/website/pull/2617 as a follow-up PR but closed in https://github.com/gitpod-io/website/pull/2617#issuecomment-1224271963 as this is already covered by the upcoming compatibility matrix.

❗ **issue**: Still placeholders remain problematic from UX perspective. 

If there's a strong sense in keeping the placeholder, let's at least update the casing and replace the token term, which is a new term for users here.

```suggestion
                        placeholder="Enter code sent via SMS"
```

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:81]: **issue**: Well, that's not so invisible over here! 🙈 

Screenshot 2022-08-23 at 7 10 16 PM

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:69]: **suggestion**: Sounds safe to keep only the top margin here.

```suggestion
                <div className="mt-2">
```

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:82]: **suggestion**: Going with hidden will also remove the occupied space here. Hopefully, this won't case any issues with referencing the classes, right?

```suggestion
                    <input type="tel" className="hidden intl-tel-input country-list" />
```

Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:82]: **issue**: Not sure which element is causing this, but would it be easy to remove the double border here?

Screenshot 2022-08-23 at 7 41 02 PM

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:59]: **issue**: Using two Greek phone number returns the following error. Not sure if I did anything wrong here. 🤷 

Screenshot 2022-08-23 at 7 42 36 PM

Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:59]: **issue**: When selecting a different country and using the sample number from the placeholder, the flow proceeds to the following screen that asks for the verification code, but still returns the same error.

Screenshot 2022-08-23 at 7 51 41 PM

**Review by gtsiolis — COMMENTED** [review]
@svenefftinge I've left another round of comments. 🥊 

I could not go through all the steps of the flow. 😭

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:59]: I had to update the preferences in twilio.

Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:59]: fixed

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:147]: **issue(non-blocking)**: The **disabled button** state should have the  "not-allowed" cursor when hovered and not relying on the CSS class for disabling the interaction. Also, relevant to the comment about the button component.

**suggestion**: Tailwind supports `cursor-not-allowed` which is something to consider when we build the button component, see [relevant docs](https://tailwindcss.com/docs/cursor).

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:148]: **suggestion**: I noticed we've skipped the _verification in progress_ step from the initial design specs in https://github.com/gitpod-io/gitpod/issues/11339#issuecomment-1183387915 which is fine, but what do you think of injecting a loading state inside the button here with the spinner icon? FWIW, we've used this pattern for running prebuilds in the past, see relevant diff below.

**thought**: We could use the same pattern also for sending the SMS via code since we rely on an external service for this action.

**thought**: The loading state of the button is another great example use case where a **button component** could be used. 💭 

https://github.com/gitpod-io/gitpod/blob/31332971069a926428151aae2eacac947ed2db21/components/dashboard/src/projects/Prebuild.tsx#L160-L169

Re-posting the verification in progress design from https://github.com/gitpod-io/gitpod/issues/11339#issuecomment-1183387915 for visibility:

> | Verification (Progress) |
> |-|
> | UserValidationModalVerificationInProgress |

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:59]: **thought**: I noticed we also skipped the resend code functionality which sounds ok for now. 🤝 

Re-posting the verification in progress design from https://github.com/gitpod-io/gitpod/issues/11339#issuecomment-1183387915 for visibility:

> | Verification (Resend) |
> |-|
> | UserValidationModalVerificationResent |

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:217]: **praise**: Thanks for giving this a go while against the [relevant discussion](https://gitpod.slack.com/archives/C01KPEPQYE6/p1657728367916349?thread_ts=1657727187.914319&cid=C01KPEPQYE6) (internal)!

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:72]: **issue(non-blocking)**: I've also run into this error but should be fine to ignore. Twilio probably hates me already. 😇 

Screenshot 2022-08-24 at 2 30 49 PM

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:215]: **suggestion**: Now that we've tried this, we probably need a new **alert component variant** for successful messages like this that is using green-ish colors. Could be fine to leave this out of the scope of these changes for now. Your call.

| Alert / Info (IMPLEMENTED) | Alert / Success (NOT IMPLEMENTED) | Alert / Success (NOT IMPLEMENTED) 🌔  |
|-|-|-|
| Frame 381 | Frame 382 | Frame 385 |

If you'd like to add this here, here's the icon (SVG) and the colors used for light and dark themes:


#### LIGHT THEME ⛅  
> Background: `green-100`
> Icon: `green-500`
> Text: `green-700`

```svg
<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M10 18C14.4183 18 18 14.4183 18 10C18 5.58172 14.4183 2 10 2C5.58172 2 2 5.58172 2 10C2 14.4183 5.58172 18 10 18ZM13.7071 8.70711C14.0976 8.31658 14.0976 7.68342 13.7071 7.29289C13.3166 6.90237 12.6834 6.90237 12.2929 7.29289L9 10.5858L7.70711 9.29289C7.31658 8.90237 6.68342 8.90237 6.29289 9.29289C5.90237 9.68342 5.90237 10.3166 6.29289 10.7071L8.29289 12.7071C8.68342 13.0976 9.31658 13.0976 9.70711 12.7071L13.7071 8.70711Z" fill="#84CC16"/>
</svg>
```

####  DARK THEME 🌔  
> Background: `green-800`
> Icon: `green-50`
> Text: `green-100`

```svg
<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M10 18C14.4183 18 18 14.4183 18 10C18 5.58172 14.4183 2 10 2C5.58172 2 2 5.58172 2 10C2 14.4183 5.58172 18 10 18ZM13.7071 8.70711C14.0976 8.31658 14.0976 7.68342 13.7071 7.29289C13.3166 6.90237 12.6834 6.90237 12.2929 7.29289L9 10.5858L7.70711 9.29289C7.31658 8.90237 6.68342 8.90237 6.29289 9.29289C5.90237 9.68342 5.90237 10.3166 6.29289 10.7071L8.29289 12.7071C8.68342 13.0976 9.31658 13.0976 9.70711 12.7071L13.7071 8.70711Z" fill="#F7FEE7"/>
</svg>
```

File: components/dashboard/src/start/VerifyModal.tsx
Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:204]: **thought**: We don't have a concrete pattern for user feedback on successful messages. So far, we've used the modal and the page for user feedback like this, see screenshots below.

| New project ✔️  | New feedback ✔️ |
|-|-|
| Screenshot 2022-08-24 at 1 15 28 PM (2) | Screenshot 2022-08-24 at 1 16 30 PM (2) |

Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:204]: **suggestion**: Given it's not ideal to update the modal title based on the user action and taking into account the [relevant discussion](https://gitpod.slack.com/archives/C01KPEPQYE6/p1657728367916349?thread_ts=1657727187.914319&cid=C01KPEPQYE6) (internal), what do you think of moving the alert component on the loading screen directly where we later on show the warnings about the latest editor release and skip asking users to acknowledge the account validation?

| Latest Release on Workspace Start | Validation on Workspace Start | 
|-|-|
| Screenshot 2022-08-24 at 2 52 07 PM (2) | Screenshot 2022-08-24 at 2 51 18 PM (2) |

Comment (gtsiolis) [inline components/dashboard/src/start/VerifyModal.tsx:204]: **thought**: In case this could be interesting we could surface the account validation status on the account settings.

**Review by gtsiolis — COMMENTED** [review]
Alright! Left a few general comments and questions below. 🔮 

Most of them can be ignored or used for opening follow up issues out of the scope of these changes. 🛹 

Thanks @svenefftinge for pushing this forward! 🏀

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:144]: That is intended. the code is uncommented so you can review the UI flow easily.

Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:144]: Ah, ok! Resolving.

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:60]: aren't the two limits you posted below what you are asking for here?

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:70]: Why would you want to limit it?

Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:70]: No strong feelings, only bringing this up because I noticed this behavior while testing with two numbers. Maybe we can leave this as is and rephrase the copy as described in https://github.com/gitpod-io/gitpod/pull/12258#discussion_r952511755? 🤷

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:217]: fixed 😇

File: components/dashboard/src/start/VerifyModal.tsx
Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:204]: I think an explicit, "you are verified" message with a 'continue' button is clearer. Especially because we are reloading the page.

Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:204]: It's a SaaS-only concern and should really touch as few places as possible. Why would it be interesting to show it there?

Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:204]: Good point! Not much interesting, indeed.

I probably had in mind how we could use the same flow to validate with a credit card if we need in the future so that users benefit from having a credit card in place when and if they enable billing for their personal account. Resolving for now. 💳

Comment (svenefftinge) [reply components/dashboard/src/start/VerifyModal.tsx:204]: We can mark users as verified when they take other actions that make them trustworthy. No UI needed.

Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:204]: Yeah, maybe also something to consider surfacing in the admin area. It will be interesting to see if users manage to bypass somehow this verification process for abusing the infrastructure. 💡 

Cc @jldec because https://github.com/gitpod-io/gitpod/issues/7597
Cc @geropl because https://github.com/gitpod-io/gitpod/issues/11469

Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:204]: **thought**: It is certainly clearer and gives users the time to acknowledge this.

**question**: What do you think of keeping modal, and updating the modal title to be the same on all cases, and rely on modal body (paragraph and alerts) to inform the user about validation errors or successful validation?

Comment (gtsiolis) [reply components/dashboard/src/start/VerifyModal.tsx:204]: Fair point. Resolving this. 🏀

**Comment by svenefftinge** [comment]
I'll hide this behind a feature flag so we can carefully enable it in prod.

**Review by corneliusludmann — APPROVED** [review]
It seems that this PR does not have changes in Team Self-Hosted owned files anymore. However, from self-hosted perspective, it looks good so far.

:bangbang: _**Hint:** You should rebase shortly before you merge this and let golden files of the installer render tests be re-generated since there are new tests in `main` already that will let fail the `main` build after merging. Sorry for the inconvenience._

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [inline components/server/src/workspace/gitpod-server-impl.ts:452]: Preferably, let's not use a field dedicated for user preferences/settings for something that actually controls the wheels or breaks. That check seems to work fine, I've tried to break without success, but I'd prefer not to have a this guarded places in a field which otherwise is free to change for users. 

There is an option that we can use to add field to tables without the downside of locking tables, we just need to add `NOT NULL DEFAULT '', ALGORITHM=INPLACE, LOCK=NONE` for the alter statement.

As we used this alternative the last time(s) discussed this matter, I'd prefer to stick to one pattern 🙏🏻

File: components/server/ee/src/billing/entitlement-service.ts
Comment (AlexTugarev) [inline components/server/ee/src/billing/entitlement-service.ts:43]: Nice integration! Fits in perfectly. 

That reminds me of the ToS flow integration, which intercepted the signup, but it would actually be better to use this kind of hook.

File: components/server/src/auth/rate-limiter.ts
Comment (AlexTugarev) [reply components/server/src/auth/rate-limiter.ts:54]: I guess this also applies to `verifyPhoneNumberVerificationToken`.

File: components/server/src/auth/verification-service.ts
Comment (AlexTugarev) [inline components/server/src/auth/verification-service.ts:72]: try-catch, otherwise it's handled as internal server error whenever one uses the happy finger on the send button.

**Review by AlexTugarev — COMMENTED** [review]
The account verification works nicely 🎉 

I left some comment, and I'm happy to discuss the `additionalData` comment 👍🏻 
Ah, the only real issue so far is that old GH account are not count as verified, because the check is not performed after signup, but rather after the following re-login.

**Comment by kylos101** [comment]
:wave:  @svenefftinge, would it be possible to include bitbucket with this too, or perhaps in a follow-on PR? I ask because of [this thread](https://gitpod.slack.com/archives/C024TUZ0MDJ/p1661457779555119).

**Comment by svenefftinge** [comment]
The logic is opt-out. So all bitbucket users need to verify.

File: components/server/src/user/user-deletion-service.ts
Comment (svenefftinge) [reply components/server/src/user/user-deletion-service.ts:70]: I intentionally left it in order to follow up in case the user did something bad.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:452]: Which discussion do you refer to?

File: components/server/src/auth/rate-limiter.ts
Comment (svenefftinge) [reply components/server/src/auth/rate-limiter.ts:54]: Ah, yes. Probably worth limiting this per user on our side, although there is a rate limit per phone number on twilio as well.

File: components/server/src/user/user-service.ts
Comment (svenefftinge) [reply components/server/src/user/user-service.ts:425]: Good catch! 🙏

File: components/server/src/auth/verification-service.ts
Comment (svenefftinge) [reply components/server/src/auth/verification-service.ts:72]: I don't get this comment. Why is there always an internal server error?

File: components/server/src/user/user-service.ts
Comment (AlexTugarev) [reply components/server/src/user/user-service.ts:425]: Let's move this check to `createUser`.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:452]: To avoid implementing the checks for nested attributes in the `additionalData` bucket we've moved the attribution override to `User` in https://github.com/gitpod-io/gitpod/pull/11522. The validation necessary is to disallow users setting the attribution target to a team id they are not a member of.

File: components/server/src/user/user-deletion-service.ts
Comment (AlexTugarev) [reply components/server/src/user/user-deletion-service.ts:70]: That we definitely need to check with data privacy. To prevent recreating accounts it would be sufficient to store a hash of the phone number, no?

File: components/server/src/auth/verification-service.ts
Comment (AlexTugarev) [reply components/server/src/auth/verification-service.ts:72]: as this is not wrapped in a `ResponseError` it'll be logged automatically as an internal server error, thus it's loud on the logs, and it seems not to be rendered properly on frontend, at least I couldn't see twillios rate limiting error rendered on frontend.

File: components/server/src/user/user-deletion-service.ts
Comment (svenefftinge) [reply components/server/src/user/user-deletion-service.ts:70]: Ok, I'll store the hash instead and introduce `d_b_blocked_phone_hashes`that we update on user block/unblock. Also will check that table on first sign up.

File: components/server/src/auth/rate-limiter.ts
Comment (svenefftinge) [reply components/server/src/auth/rate-limiter.ts:54]: I'm introducing a new category `phoneVerification` with 10 points per 10 seconds.

File: components/server/src/user/user-deletion-service.ts
Comment (svenefftinge) [reply components/server/src/user/user-deletion-service.ts:70]: Update: I'm leaving the blocked_phone_hashes out of this PR, as it makes it even bigger (it already is big).
But I'm moving the two properties into `d_b_user`.

File: components/server/src/auth/verification-service.ts
Comment (svenefftinge) [reply components/server/src/auth/verification-service.ts:72]: It is fine to be loud on the logs. It's not a user error when an error is thrown but something with the system. The error is thrown to frontend and I'll make it so that the user gets feedback about the error.

File: components/server/src/user/user-service.ts
Comment (svenefftinge) [reply components/server/src/user/user-service.ts]: moved it to the callback that is passed to `createUser` because I needed access to the authUser object. Also looked like a good pace for it otherwise. Fine?

**Comment by jldec** [comment]
/werft run with-clean-slate-deployment=true
🤞

File: components/server/src/user/user-deletion-service.ts
Comment (AlexTugarev) [inline components/server/src/user/user-deletion-service.ts:153]: `ER_DATA_TOO_LONG: Data too long for column 'verificationPhoneNumber'`

we're almost there, just 2 chars too long.

Comment (svenefftinge) [reply components/server/src/user/user-deletion-service.ts:153]: what did you type in?

Comment (svenefftinge) [reply components/server/src/user/user-deletion-service.ts:153]: ah, got it.

Comment (AlexTugarev) [reply components/server/src/user/user-deletion-service.ts]: That happens on deletion of an account, which actually breaks the process.

Comment (AlexTugarev) [reply components/server/src/user/user-deletion-service.ts]: `deleted-phoneNumber` with a length of 20 should work just fine.

**Review by AlexTugarev — APPROVED** [review]
LGTM!
Tested works as advertised.

Logic for `Test 1 ++` isn't there is code, which I think is not needed at all ;-)

**Pull request merged**
